service/hub/server/server.go

```diff
@@ -0,0 +1,355 @@
+)
+
+var (
+	hostPortRegex = regexp.MustCompile(`^(?:\[(.+)\]:\d+|([^:]+):\d+)$`)
+	scopeRegex    = regexp.MustCompile(`([a-z0-9]+)(\([a-z0-9]+\))?`)
+)
+
+type AuthServer struct {
+	config         *Config
+	authenticators api.Authenticator
+	authorizers    api.Authorizer
+}
+
+func NewAuthServer(c *Config) (*AuthServer, error) {
+	as := &AuthServer{
+		config:         c,
+		authenticators: auth.NewSealosAuthn(),
+		authorizers:    auth.NewSealosAuthz(),
+	}
+	return as, nil
+}
+
+type AuthRequest struct {
+	RemoteConnAddr string
+	RemoteAddr     string
+	RemoteIP       net.IP
+	User           string
+	Password       api.PasswordString
+	Account        string
+	Service        string
+	Scopes         []AuthScope
+	Labels         api.Labels
+}
+
+type AuthScope struct {
+	Type    string
+	Class   string
+	Name    string
+	Actions []string
+}
+
+type AuthzResult struct {
+	scope            AuthScope
+	autorizedActions []string
+}
+
+func (ar AuthRequest) String() string {
+	return fmt.Sprintf("{%s:%s@%s %s}", ar.User, ar.Password, ar.RemoteAddr, ar.Scopes)
+}
+
+func parseRemoteAddr(ra string) net.IP {
+	hp := hostPortRegex.FindStringSubmatch(ra)
+	if hp != nil {
+		if hp[1] != "" {
+			ra = hp[1]
+		} else if hp[2] != "" {
+			ra = hp[2]
+		}
+	}
+	res := net.ParseIP(ra)
+	return res
+}
+
+func parseScope(scope string) (string, string, error) {
+	parts := scopeRegex.FindStringSubmatch(scope)
+	if parts == nil {
+		return "", "", fmt.Errorf("malformed scope request")
+	}
+
+	switch len(parts) {
+	case 3:
+		return parts[1], "", nil
+	case 4:
+		return parts[1], parts[3], nil
+	default:
+		return "", "", fmt.Errorf("malformed scope request")
+	}
+}
+
+func (as *AuthServer) ParseRequest(req *http.Request) (*AuthRequest, error) {
+	ar := &AuthRequest{RemoteConnAddr: req.RemoteAddr, RemoteAddr: req.RemoteAddr}
+	ar.RemoteIP = parseRemoteAddr(ar.RemoteAddr)
+	if ar.RemoteIP == nil {
+		return nil, fmt.Errorf("unable to parse remote addr %s", ar.RemoteAddr)
+	}
+	user, password, haveBasicAuth := req.BasicAuth()
+	if haveBasicAuth {
+		ar.User = user
+		ar.Password = api.PasswordString(password)
+	} else if req.Method == "POST" {
+		// username and password could be part of form data
+		username := req.FormValue("username")
+		password := req.FormValue("password")
+		if username != "" && password != "" {
+			ar.User = username
+			ar.Password = api.PasswordString(password)
+		}
+	}
+	ar.Account = req.FormValue("account")
+	if ar.Account == "" {
+		ar.Account = ar.User
+	} else if haveBasicAuth && ar.Account != ar.User {
+		return nil, fmt.Errorf("user and account are not the same (%q vs %q)", ar.User, ar.Account)
+	}
+	ar.Service = req.FormValue("service")
+	if err := req.ParseForm(); err != nil {
+		return nil, fmt.Errorf("invalid form value")
+	}
+	// https://github.com/docker/distribution/blob/1b9ab303a477ded9bdd3fc97e9119fa8f9e58fca/docs/spec/auth/scope.md#resource-scope-grammar
+	if req.FormValue("scope") != "" {
```

Comment:
https://pkg.go.dev/github.com/docker/distribution/reference?utm_source=godoc seems to use this lib is a better way.